lib/float/common.sail

```diff
@@ -0,0 +1,78 @@
+/* Copyright 2024 Intel Corporation                                         */
+/*   Pan Li - pan2.li@intel.com                                             */
+/*                                                                          */
+/* Redistribution and use in source and binary forms, with or without       */
+/* modification, are permitted provided that the following conditions are   */
+/* met:                                                                     */
+/*                                                                          */
+/* 1. Redistributions of source code must retain the above copyright        */
+/*    notice, this list of conditions and the following disclaimer.         */
+/* 2. Redistributions in binary form must reproduce the above copyright     */
+/*    notice, this list of conditions and the following disclaimer in the   */
+/*    documentation and/or other materials provided with the distribution.  */
+/*                                                                          */
+/* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS      */
+/* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT        */
+/* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A  */
+/* PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT       */
+/* HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,   */
+/* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED */
+/* TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR   */
+/* PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF   */
+/* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING     */
+/* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS       */
+/* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.             */
+/*==========================================================================*/
+
+$ifndef _FLOAT_COMMON
+$define _FLOAT_COMMON
+
+/* Floating point types definition */
+type fp_exception_flags = bits(5) /* Floating-point exception flags. */
+type fp_bits = { 'n, 'n in {16, 32, 64}. bits('n) } /* Floating-point in bits. */
+
+/* Floating point constants */
+let fp_eflag_invalid : fp_exception_flags = 0b00001
+
+/* Floating point struct */
+$[undefined_gen forbid] /* Todo: temporarlly workaround a type check error. */
+struct float_bits('n : Int) = {
+  sign     : bits(1),
+  exp      : bits(if  'n == 16
+                  then 5
+                  else (if   'n == 32
+                        then 8
+                        else 11)),
+  mantissa : bits(if  'n == 16
+                  then 10
+                  else (if   'n == 32
+                        then 23
+                        else 52)),
+}
+
+/* The val func implementations */
+val      float_decompose : forall 'n, 'n in { 16, 32, 64 }. bits('n) -> float_bits('n)
```

Comment:
HI @Alasdair

Move the question to PR, it looks like inline comments are not easy to locate in the file.

I am trying to prepare the float_compose that convert struct float_bits('n) to bits('n), but I am not sure if there is a better approach. The below implement may work but looks "silly" up to a point. I try to do something like `op[15..15] = fp.sign` but failed to find a sample for it.

```
  92 + │ val      float_compose : forall 'n, 'n in { 16, 32, 64, 128 }. float_bits('n) -> bits('n)
  93 + │ function float_compose(fp_bits) = {
  94 + │   let sign     = sail_zero_extend (fp_bits.sign, 'n);
  95 + │   let exp      = sail_zero_extend (fp_bits.exp, 'n);
  96 + │   let mantissa = sail_zero_extend (fp_bits.mantissa, 'n);
  97 + │   match 'n {
  98 + │     16  => add_bits (sign << 15,  exp << 10,  mantissa),
  99 + │     32  => add_bits (sign << 31,  exp << 23,  mantissa),
 100 + │     64  => add_bits (sign << 63,  exp << 52,  mantissa),
 101 + │     128 => add_bits (sign << 127, exp << 112, mantissa),
 102 + │   }
 103 + │ }
```

It may also have some error like below:
```
94 |  let sign     = sail_zero_extend (fp_bits.sign, 'n);
                                                                            ^^ checking function argument has type int('m)
   Cannot re-write sizeof('n)
```

Does that mean we need to convert for `'n` here? 
Thanks a lot.

Reply:
Is there any suggestion here? Thanks a lot.